File: JavaScript/randomized_quick_sort.js
function randomizedQuickSort(arr, low = 0, high = arr.length - 1) {
    if (low < high) {
        const pivotIndex = randomizedPartition(arr, low, high);
        randomizedQuickSort(arr, low, pivotIndex - 1);
        randomizedQuickSort(arr, pivotIndex + 1, high);
    }
    return arr;
}

function randomizedPartition(arr, low, high) {
    const randomIndex = Math.floor(Math.random() * (high - low + 1)) + low;
    [arr[randomIndex], arr[high]] = [arr[high], arr[randomIndex]];
    return partition(arr, low, high);
}

function partition(arr, low, high) {
    const pivot = arr[high];
    let i = low - 1;
    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
    return i + 1;
}

const array = [10, 7, 8, 9, 1, 5, 0];
console.log("Sorted array:", randomizedQuickSort(array));

File: Kotlin/randomizedQuickSort.kt
import kotlin.random.Random

fun randomizedQuickSort(arr: IntArray, low: Int, high: Int) {
    if (low < high) {
        val pivotIndex = randomizedPartition(arr, low, high)
        randomizedQuickSort(arr, low, pivotIndex - 1)
        randomizedQuickSort(arr, pivotIndex + 1, high)
    }
}

fun randomizedPartition(arr: IntArray, low: Int, high: Int): Int {
    val randomIndex = Random.nextInt(low, high + 1)
    arr.swap(randomIndex, high)
    return partition(arr, low, high)
}

fun partition(arr: IntArray, low: Int, high: Int): Int {
    val pivot = arr[high]
    var i = low - 1
    for (j in low until high) {
        if (arr[j] <= pivot) {
            i++
            arr.swap(i, j)
        }
    }
    arr.swap(i + 1, high)
    return i + 1
}

fun IntArray.swap(i: Int, j: Int) {
    val temp = this[i]
    this[i] = this[j]
    this[j] = temp
}

fun main() {
    val arr = intArrayOf(10, 7, 8, 9, 1, 5, 0, -1, -5)
    randomizedQuickSort(arr, 0, arr.size - 1)
    println("Sorted array: ${arr.joinToString(", ")}")
}